sbtc/src/testing/containers.rs

```diff
@@ -0,0 +1,220 @@
+    format!(
+        "{}/../docker/tests/{file_name}",
+        std::env::var("CARGO_MANIFEST_DIR").unwrap()
+    )
+}
+
+/// Builder for test containers
+pub struct TestContainersBuilder {
+    /// Compose files to use
+    compose_files: Vec<String>,
+}
+
+impl TestContainersBuilder {
+    /// Create a new `TestContainersBuilder`
+    pub fn new() -> Self {
+        Self { compose_files: vec![] }
+    }
+    /// Add compose file
+    fn with_compose(mut self, filename: &str) -> Self {
+        self.compose_files.push(filename.to_string());
+        self
+    }
+    /// Build a new `TestContainers` with this config
+    pub fn build(self) -> TestContainers {
+        TestContainers::new(self)
+    }
+    /// Build a new `TestContainers` with this config and start it, panic if fails
+    async fn start(self) -> TestContainers {
+        let mut stack = self.build();
+        stack.up().await.expect("failed to start the stack");
+        stack
+    }
+    /// Require Bitcoin service
+    pub fn with_bitcoin(self) -> Self {
+        self.with_compose(&compose_path(COMPOSE_BITCOIN))
+    }
+    /// Start the test stack with only Bitcoin, panic if fails
+    pub async fn start_bitcoin() -> TestContainers {
+        Self::new().with_bitcoin().start().await
+    }
+}
+
+impl Default for TestContainersBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A `TestContainers` manage an isolated docker stack for tests
+pub struct TestContainers {
+    /// The underlying compose stack
+    compose: ManuallyDrop<DockerCompose>,
+    /// By default the stack is downed on drop (triggered on test exit)
+    down_on_drop: bool,
+    /// The Bitcoin container, if present
+    bitcoin: OnceCell<BitcoinContainer>,
+}
+impl TestContainers {
+    /// Create a new `TestContainers`
+    pub fn new(config: TestContainersBuilder) -> Self {
+        let compose = DockerCompose::with_local_client(&config.compose_files);
+        Self {
+            compose: ManuallyDrop::new(compose),
+            down_on_drop: true,
+            bitcoin: OnceCell::new(),
+        }
+    }
+
+    /// Up the required services
+    pub async fn up(&mut self) -> Result<(), Error> {
+        self.compose.up().await.map_err(Error::ComposeError)
+    }
+
+    /// Get the service exposed port given the internal port
+    pub async fn get_service_port(&self, service: &str, internal_port: u16) -> Result<u16, Error> {
+        let container = self.compose.service(service).ok_or(Error::ComposeError(
+            testcontainers::compose::ComposeError::ServiceNotFound(service.to_string()),
+        ))?;
+        container
+            .get_host_port_ipv4(internal_port)
+            .await
+            .map_err(Error::Testcontainers)
+    }
+
+    /// Get the service host
+    pub async fn get_service_host(&self, service: &str) -> Result<url::Host, Error> {
+        let container = self.compose.service(service).ok_or(Error::ComposeError(
```

Comment:
small nit: we should use `ok_or_else` for these things. This is test code so I don't actually care, just calling it out.